stablehlo/reference/Ops.cpp

```diff
@@ -702,6 +723,51 @@ Token evalAfterAllOp(ArrayRef<Token> inputs, MLIRContext *context) {
   return Token(context);
 }
 
+Tensor evalAllReduceOp(const Tensor &operand,
+                       SmallVector<SmallVector<uint32_t>> replicaGroups,
+                       int64_t channelId, bool useGlobalDeviceIds,
+                       Region &region, Process *process, Scope &scope,
+                       ShapedType resultType) {
+  if (!process)
+    llvm::report_fatal_error(
+        "all_reduce is only supported when run via interpreter.run_parallel");
+
+  ProcessGroups processGroups;
+  if (channelId <= 0 && !useGlobalDeviceIds)
+    processGroups = process->crossReplica(replicaGroups);
+  if (channelId > 0 && !useGlobalDeviceIds)
+    processGroups = process->crossReplicaAndPartition(replicaGroups);
+  if (channelId > 0 && useGlobalDeviceIds)
+    processGroups = process->flattenedIds(replicaGroups);
+
+  auto processGroup = processGroups.findAll(process->getId())[0];
+  auto groupOperands =
+      process->rendezvous(processGroup, channelId, operand).getTensors();
+
+  Tensor result(resultType);
+  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
+       ++resultIt) {
+    SmallVector<Tensor> inputs;
+    for (auto input : groupOperands)
+      inputs.push_back(
+          Tensor(RankedTensorType::get({1}, operand.getElementType()),
+                 input.get(*resultIt)));
+
+    auto concatResultShape = inputs[0].getShape();
+    concatResultShape[0] = inputs.size();
+    auto concatenateResult = evalConcatenateOp(
+        inputs, 0,
+        RankedTensorType::get(concatResultShape, operand.getElementType()));
+
+    result.set(*resultIt,
+               evalReduceOp(concatenateResult,
+                            Tensor(convert(operand.getElementType(), 0.0)), {0},
+                            region, process, scope)[0]
+                   .get({}));
+  }
```

Comment:
I think this implementation complicates things quite a bit. The following implementation that directly follows the spec seems pretty straightforward:

```C++
Tensor result(resultType);
for (auto resultIt = result.index_begin(); resultIt != result.index_end();
     ++resultIt) {
  Tensor resultElement;
  for (auto groupOperand : groupOperands) {
    Tensor groupOperandElement(
        RankedTensorType::get({}, groupOperand.getElementType()),
        groupOperand.get(*resultIt));
    if (resultElement) {
      resultElement = eval(computation, {resultElement, groupOperandElement},
                           process, &scope)[0]
                          .getTensor();
    } else {
      resultElement = groupOperandElement;
    }
  }
  result.set(*resultIt, resultElement.get({}));
}
```